### PR TITLE
Build mobile inference library for minimum size

### LINKED ISSUE #1845: build for minimum size
## Paddle bin
当前通过CMake最小编译出来的Paddle预测程序大小是7.38M。其中预测程序链接时依赖的库如下。

| Library | size |
|---------|-----|
| libpaddle_cuda.a | 30496 |
| libpaddle_function.a | 718998 |
| libpaddle_gserver.a | 6753326 |
| libpaddle_network.a | 118668 |
| libpaddle_pserver.a | 1192704 |
| libpaddle_trainer_lib.a | 915834 |
| libpaddle_math.a | 3100042 |
| libpaddle_utils.a | 279696 |
| libpaddle_parameter.a | 1009290 |
| libpaddle_api.a | 349126 |
| libpaddle_proto.a | 1549498 |
| libgflags.a | 197254 |
| libglog.a | 189138 |
| libz.a | 112010 |
| libprotobuf.a | 4323004 |
| libopenblas.a | 3279128 |

## 减小Paddle bin大小的方法
1. 预测程序去掉对libpaddle_trainer_lib.a, libpaddle_pserver.a和libpaddle_network.a的依赖；
2. 预测程序去掉对libpaddle_function.a和libpaddle_gserver.a的whole archive；
3. 其他Library szie分析;
4. strip等方法

### PR DESCRIPTION
**Pull request by hedaoyuan**
fix #1845 
- [x] Remove pserver and trainer
- [x] Remove evaluators
- [x] Remove some useless gradientmachines
- [x] Remove dataproviders
- [x] Remove optimizer

------------------------------------------------------------------------------
- Before optimize

| Library | size(kB) |
| ----------------- | ---------------- |
| libpaddle_capi_shared.so | 9185 |
| libpaddle_capi_shared.so(strip) | 4212 |
| ibpaddle_capi_whole.a | 19641 |

- After optimize

| Library | size(kB) |
| ----------------- | ---------------- |
| libpaddle_capi_shared.so | 8501 |
| libpaddle_capi_shared.so(strip) | 3539 |
| ibpaddle_capi_whole.a | 15904 |

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -86,6 +86,14 @@ if(ANDROID OR IOS)
         "Disable MKLDNN when cross-compiling for Android and iOS" FORCE)
     set(WITH_MKLML OFF CACHE STRING
         "Disable MKLML package when cross-compiling for Android and iOS" FORCE)
+
+    if(WITH_C_API)
+      # Compile PaddlePaddle mobile inference library
+      set(MOBILE_INFERENCE ON)
+      add_definitions(-DPADDLE_MOBILE_INFERENCE)
+    endif()
+    set(WITH_TESTING OFF CACHE STRING "Disable TESTING when cross-compiling
```
Comment (Xreki) [inline CMakeLists.txt]: I'd like to not set `WITH_TESTING` to `OFF`, because it is useful to check the linking error.

File: CMakeLists.txt [hunk 2]
```diff
@@ -86,6 +86,14 @@ if(ANDROID OR IOS)
         "Disable MKLDNN when cross-compiling for Android and iOS" FORCE)
     set(WITH_MKLML OFF CACHE STRING
         "Disable MKLML package when cross-compiling for Android and iOS" FORCE)
+
+    if(WITH_C_API)
```
Comment (Xreki) [inline CMakeLists.txt]: Do we need to always set `WITH_C_API` to `ON`?

File: cmake/util.cmake [hunk 3]
```diff
@@ -73,25 +73,43 @@ function(link_paddle_exe TARGET_NAME)
...
+            paddle_parameter
+            paddle_proto
+            paddle_cuda
+            paddle_optimizer
+            ${EXTERNAL_LIBS}
+            ${CMAKE_THREAD_LIBS_INIT}
+            ${CMAKE_DL_LIBS}
+            ${RDMA_LD_FLAGS}
+            ${RDMA_LIBS})
+    endif()
```
Comment (Xreki) [inline cmake/util.cmake]: Maybe we can use a new variable to specify which modules need to link, like:
```
if(MOBILE_INFERENCE)
    set(PADDLE_LIBS
        paddle_math
        paddle_utils
        paddle_parameter
        paddle_proto
        paddle_cuda
        )
else()
    set(PADDLE_LIBS
        paddle_pserver
        paddle_trainer_lib
        paddle_network
        paddle_math
        paddle_utils
        paddle_parameter
        paddle_proto
        paddle_cuda
        paddle_optimizer         
        )
endif()
target_circle_link_libraries(${TARGET_NAME}
        ARCHIVE_START
        paddle_gserver
        paddle_function
        ARCHIVE_END
        ${PADDLE_LIBS}
        ${EXTERNAL_LIBS}
        ${CMAKE_THREAD_LIBS_INIT}
        ${CMAKE_DL_LIBS}
        ${RDMA_LD_FLAGS}
        ${RDMA_LIBS})
 endif()

```

File: paddle/capi/CMakeLists.txt [hunk 4]
```diff
@@ -50,7 +48,9 @@ if(NOT IOS)
   add_library(paddle_capi_shared SHARED ${CAPI_SOURCES})
   set_target_properties(paddle_capi_shared	PROPERTIES LINK_FLAGS "${LINK_FLAGS}")
   target_include_directories(paddle_capi_shared PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
-  link_paddle_exe(paddle_capi_shared)
+
+link_paddle_exe(paddle_capi_shared)
```
Comment (Xreki) [inline paddle/capi/CMakeLists.txt]: Need indentations here.

File: paddle/gserver/CMakeLists.txt [hunk 5]
```diff
@@ -60,6 +60,36 @@ if(NOT WITH_PYTHON)
...
+    list(REMOVE_ITEM GSERVER_SOURCES
+         gradientmachines/MultiNetwork.cpp
+         gradientmachines/RecurrentGradientMachine.cpp
+         gradientmachines/ParallelNeuralNetwork.cpp
+         gradientmachines/GradientMachineMode.cpp
+         gradientmachines/MultiGradientMachine.cpp)
+
+    # Remove useless layers
+    list(REMOVE_ITEM GSERVER_SOURCES
+    	 layers/RecurrentLayerGroup.cpp)
```
Comment (Xreki) [inline paddle/gserver/CMakeLists.txt]: I am not sure whether `layers/RecurrentLayerGroup.cpp` can be removed.
And cost layers can be removed too?

Comment (hedaoyuan) [reply paddle/gserver/CMakeLists.txt]: For `RecurrentLayerGroup.cpp`, first remove `Evaluator`, then remove the `RecurrentGradientMachine` which depends on `Evaluator`, and last remove 'RecurrentLayerGroup' which depends on 'RecurrentGradientMachine'.

For `cost layers`, can be considered in the next PR.

File: CMakeLists.txt (hunk 2 above)
Comment (hedaoyuan) [reply CMakeLists.txt]: Done.

File: paddle/capi/CMakeLists.txt (hunk 4 above)
Comment (hedaoyuan) [reply paddle/capi/CMakeLists.txt]: Done

File: CMakeLists.txt (hunk 1 above)
Comment (hedaoyuan) [reply CMakeLists.txt]: Done.

File: cmake/util.cmake (hunk 3 above)
Comment (hedaoyuan) [reply cmake/util.cmake]: I will fix this commit in the next pr.

**Comment by NHZlX** [comment]
Have tested, LGTM

**Pull request merged**